[CS] Handle holes in `mergeEquivalenceClasses`

hamishknight opened the pull request:
Make sure we preserve `TVO_CanBindToHole`. This unfortunately requires introducing the concept of a "non representative" hole for representatives where the actual hole is given by another type variable in the equivalence class. This avoids disturbing binding order and regressing diagnostics. Once binding is less sensitive to ordering we ought to be able to always choose the hole as the representative.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented on test/Constraints/diagnostics.swift:
```diff
@@ -1578,3 +1578,13 @@ func testAddMemberVsRemoveCall() {
+  let x = foo([], i)
+  // expected-error@-1 {{generic parameter 'T' could not be inferred}}
+  // expected-error@-2 {{cannot convert value of type '[Any]' to expected argument type '_.Type'}}
+  // FIXME: The above should say `T.Type` not `_.Type` ^
```
I think this is a bug in `FailureDiagnostic::resolveType`, might be easy to fix.

hamishknight replied:
Yeah the issue is that we need the hole's locator, but that currently requires iterating over the equivalence class, which we don't have in the Solution. So either we need to stick the locator on the type variable itself, or stick a mapping in the Solution

xedin replied:
Oh, I see.

hamishknight replied:
The actual type being diagnosed is part of the fix though, right? Unless I'm misunderstanding what you mean by transformation

xedin replied:
We never want to print type variables so maybe replacing them with an underlying generic parameter should be done as part of forming a solution instead of delaying it to diagnostics like we do right now.

hamishknight commented:
@swift-ci please test

hamishknight commented on validation-test/Sema/SwiftUI/rdar88256059.swift:
```diff
@@ -9,7 +9,8 @@ struct MyView: View {
     Table(self.data) {
-      // expected-error@-1 {{expected expression of type 'Columns' in result builder 'TableColumnBuilder'}} {{23-23=<#T##Columns#>}}
+      // expected-error@-1 {{expected expression of type 'Column' in result builder 'TableColumnBuilder'}} {{23-23=<#T##Column#>}}
+      // expected-error@-2 {{generic parameter 'Column' could not be inferred}}
```
This is a little unfortunate, but it's because we're now no longer stripping `TVO_CanBindToHole` from the generic parameter. We can't eagerly bind the missing argument to a hole because we rely on type inference to infer the type for better diagnostics. Maybe we could rework those diagnostics to instead look at the parameter type, or maybe we could tweak the diagnostic logic for a defaulted generic parameter to account for missing argument fixes (unfortunately it currently gets defaulted before we match arguments though). In any case, I'm going to leave this as future work.

hamishknight committed:
[CS] Handle holes in `mergeEquivalenceClasses`

Make sure we preserve `TVO_CanBindToHole`. This unfortunately requires
introducing the concept of a "non-representative" hole for
representatives where the actual hole is given by another type variable
in the equivalence class. This avoids disturbing binding order and
regressing diagnostics. Once binding is less sensitive to ordering we
ought to be able to always choose the hole as the representative.

hamishknight committed:
[CS] Improve applicable fn handling of hole function type

Unwrap a metatype before checking if we have an associated fix.

hamishknight committed:
[CS] NFC: Split out locator checks for `isClosure(Parameter|Result)Type`

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please SourceKit stress test
